### PR TITLE
Feature/sg optional progressbar prediction

### PR DESCRIPTION
**Pull request by EduardoPach**
What does this PR do:

Adds functionality mentioned in https://github.com/Deci-AI/super-gradients/issues/1329

### CODE REVIEW

**Comment by EduardoPach** [comment]
@Louis-Dupont can you review it?

File: src/super_gradients/training/models/detection_models/yolo_base.py [hunk 1]
```diff
@@ -570,9 +571,10 @@ def predict(
                             If None, the default value associated to the training is used.
         :param batch_size:  Maximum number of images to process at the same time.
         :param fuse_model:  If True, create a copy of the model, and fuse some of its layers to increase performance. This increases memory usage.
+        :param show_progress_bar: If True, show a progress bar.
```
Comment (BloodAxe) [inline src/super_gradients/training/models/detection_models/yolo_base.py:574]: Please ensure indentation is preserved:
```
        :param batch_size:        Maximum number of images to process at the same time.
        :param fuse_model:        If True, create a copy of the model, and fuse some of its layers to increase performance. This increases memory usage.
        :param show_progress_bar: If True, show a progress bar.
```

File: src/super_gradients/training/pipelines/pipelines.py [hunk 2]
```diff
@@ -90,7 +90,9 @@ def _fuse_model(self, input_example: torch.Tensor):
         self.model.prep_model_for_conversion(input_size=input_example.shape[-2:])
         self.fuse_model = False
 
-    def __call__(self, inputs: Union[str, ImageSource, List[ImageSource]], batch_size: Optional[int] = 32) -> ImagesPredictions:
+    def __call__(
+        self, inputs: Union[str, ImageSource, List[ImageSource]], batch_size: Optional[int] = 32, show_progress_bar: Optional[bool] = True
```
Comment (BloodAxe) [inline src/super_gradients/training/pipelines/pipelines.py:94]: No need for Optional[] here.

File: src/super_gradients/training/pipelines/pipelines.py [hunk 3]
```diff
@@ -102,39 +104,44 @@ def __call__(self, inputs: Union[str, ImageSource, List[ImageSource]], batch_siz
 
         :param inputs:      inputs to the model, which can be any of the above-mentioned types.
         :param batch_size:  Maximum number of images to process at the same time.
+        :param show_progress_bar:   If True, show a progress bar.
         :return:            Results of the prediction.
         """
 
         if includes_video_extension(inputs):
-            return self.predict_video(inputs, batch_size)
+            return self.predict_video(inputs, batch_size, show_progress_bar)
```
Comment (BloodAxe) [inline src/super_gradients/training/pipelines/pipelines.py:112]: Please use explicit arguments here `show_progress_bar=show_progress_bar` and elsewhere

**Comment by BloodAxe** [comment]
Thanks for your PR. Unfortunately it look like it contains **unsigned** commits with is a hard requirement. Please check https://github.com/Deci-AI/super-gradients/blob/master/CONTRIBUTING.md for information on setting up the signed commits.

**Comment by BloodAxe** [comment]
No. What you need to do is create another branch starting from master, then move your commits from PR branch to the 
new branch and force push it to the remote under the old name. Let's say your PR branch named 'feature/my_awesome_pr'
then you need to do the following:

```bash
git checkout master
git checkout -b feature/my_awesome_pr_signed
git merge --squash --no-commit feature/my_awesome_pr
git commit -S -m "My signed commit"
git push -f origin feature/my_awesome_pr_signed:feature/my_awesome_pr
```

Last command will overwrite your PR branch with the new signed commit containing all changes from the PR. 

LMK if this snippet would work (written from memory). If yes, we will update contribution guide to contain this text to help other contributors (Btw, if it works feel free to make a second PR with these instructions).

**Comment by EduardoPach** [comment]
> No. What you need to do is create another branch starting from master, then move your commits from PR branch to the new branch and force push it to the remote under the old name. Let's say your PR branch named 'feature/my_awesome_pr' then you need to do the following:
> 
> ```shell
> git checkout master
> git checkout -b feature/my_awesome_pr_signed
> git merge --squash --no-commit feature/my_awesome_pr
> git commit -S -m "My signed commit"
> git push -f origin feature/my_awesome_pr_signed:feature/my_awesome_pr
> ```
> 
> Last command will overwrite your PR branch with the new signed commit containing all changes from the PR.
> 
> LMK if this snippet would work (written from memory). If yes, we will update contribution guide to contain this text to help other contributors (Btw, if it works feel free to make a second PR with these instructions).

It seemed to work! I'll update the `CONTRIBUTING.md` shortly

**Comment by BloodAxe** [comment]
Glad it helped. Regarding PR to contributing md - probably no need, there is already one made by me https://github.com/Deci-AI/super-gradients/pull/1486 :)

File: src/super_gradients/training/pipelines/pipelines.py [hunk 4]
```diff
@@ -436,9 +451,9 @@ def _instantiate_image_prediction(self, image: np.ndarray, prediction: Classific
         return ImageClassificationPrediction(image=image, prediction=prediction, class_names=self.class_names)
 
     def _combine_image_prediction_to_images(
-        self, images_predictions: Iterable[ImageClassificationPrediction], n_images: Optional[int] = None
+        self, images_predictions: Iterable[ImageClassificationPrediction], n_images: Optional[int] = None, show_progress_bar: bool = True
     ) -> ImagesClassificationPrediction:
-        if n_images is not None and n_images == 1:
+        if (n_images is not None and n_images == 1) or not show_progress_bar:
```
Comment (BloodAxe) [inline src/super_gradients/training/pipelines/pipelines.py:456]: So here you've introduced a bug in my opinion.
By adding `or not show_progress_bar:` to the condition `if (n_images is not None and n_images == 1)` you are forcing code to go `images_predictions = [next(iter(images_predictions))]` path if progress bar is disabled.
But this path would process only one element of the sequence, not all of them.
What you should be doing instead is this:
`images_predictions = [image_predictions for image_predictions in tqdm(images_predictions, total=n_images, desc="Predicting Images", disable=not show_progress_bar)]`

Btw it is a good idea to add a test to `tests/unit_tests/test_predict.py` which would take any detection model and run predict on few images with show_progress_bar=False and ensure than number of predictions matches number of input images.

**Review by BloodAxe — CHANGES_REQUESTED** [review]
https://github.com/Deci-AI/super-gradients/pull/1465#discussion_r1338094118

**Comment by EduardoPach** [comment]
> [#1465 (comment)](https://github.com/Deci-AI/super-gradients/pull/1465#discussion_r1338094118)

Sorry for the delay, gonna fix this right now

File: src/super_gradients/training/pipelines/pipelines.py [hunk 5]
```diff
@@ -304,20 +313,25 @@ def _instantiate_image_prediction(self, image: np.ndarray, prediction: Detection
...
 
     def _combine_image_prediction_to_video(
-        self, images_predictions: Iterable[ImageDetectionPrediction], fps: float, n_images: Optional[int] = None
+        self, images_predictions: Iterable[ImageDetectionPrediction], fps: float, n_images: Optional[int] = None, show_progress_bar: bool = True
     ) -> VideoDetectionPrediction:
-        images_predictions = [image_predictions for image_predictions in tqdm(images_predictions, total=n_images, desc="Predicting Video")]
+        if not show_progress_bar:
+            images_predictions = [next(iter(images_predictions))]
+        else:
+            images_predictions = [image_predictions for image_predictions in tqdm(images_predictions, total=n_images, desc="Predicting Video")]
```
Comment (BloodAxe) [inline src/super_gradients/training/pipelines/pipelines.py:334]: Here `disable=not show_progress_bar` is missing

File: src/super_gradients/training/pipelines/pipelines.py [hunk 6]
```diff
@@ -304,20 +313,25 @@ def _instantiate_image_prediction(self, image: np.ndarray, prediction: Detection
...
+            ]
 
         return ImagesDetectionPrediction(_images_prediction_lst=images_predictions)
 
     def _combine_image_prediction_to_video(
-        self, images_predictions: Iterable[ImageDetectionPrediction], fps: float, n_images: Optional[int] = None
+        self, images_predictions: Iterable[ImageDetectionPrediction], fps: float, n_images: Optional[int] = None, show_progress_bar: bool = True
     ) -> VideoDetectionPrediction:
-        images_predictions = [image_predictions for image_predictions in tqdm(images_predictions, total=n_images, desc="Predicting Video")]
+        if not show_progress_bar:
```
Comment (BloodAxe) [inline src/super_gradients/training/pipelines/pipelines.py:331]: And this check does not make sense to me. Why are you changing logic of original implementation?

File: src/super_gradients/training/pipelines/pipelines.py [hunk 7]
```diff
@@ -377,20 +391,25 @@ def _instantiate_image_prediction(self, image: np.ndarray, prediction: PoseEstim
...
+            ]
 
         return ImagesPoseEstimationPrediction(_images_prediction_lst=images_predictions)
 
     def _combine_image_prediction_to_video(
-        self, images_predictions: Iterable[ImageDetectionPrediction], fps: float, n_images: Optional[int] = None
+        self, images_predictions: Iterable[ImageDetectionPrediction], fps: float, n_images: Optional[int] = None, show_progress_bar: bool = True
     ) -> VideoPoseEstimationPrediction:
-        images_predictions = [image_predictions for image_predictions in tqdm(images_predictions, total=n_images, desc="Predicting Video")]
+        if not show_progress_bar:
```
Comment (BloodAxe) [inline src/super_gradients/training/pipelines/pipelines.py:409]: Same here. Do not add IFs where they were not present.

File: src/super_gradients/training/pipelines/pipelines.py [hunk 8]
```diff
@@ -377,20 +391,25 @@ def _instantiate_image_prediction(self, image: np.ndarray, prediction: PoseEstim
...
+            images_predictions = [next(iter(images_predictions))]
+        else:
+            images_predictions = [image_predictions for image_predictions in tqdm(images_predictions, total=n_images, desc="Predicting Video")]
```
Comment (BloodAxe) [inline src/super_gradients/training/pipelines/pipelines.py:412]: `disable` is missing

**Comment by BloodAxe** [comment]
@EduardoPach there is still some leftovers remaining. I left some comments for you to address.

**Comment by EduardoPach** [comment]
@BloodAxe Now I believe should be done!

**Review by BloodAxe — APPROVED** [review]
LGTM

**Comment by BloodAxe** [comment]
@EduardoPach I think we are ready to merge your PR, however i still see some non-signed commits present in this PR, like this [one](https://github.com/Deci-AI/super-gradients/pull/1465/commits/6d61443f3d3b20bef16aa7fc26315640f6d7e534).
Can you please fix this and we can finally merge it )

**Comment by EduardoPach** [comment]
> @EduardoPach I think we are ready to merge your PR, however i still see some non-signed commits present in this PR, like this [one](https://github.com/Deci-AI/super-gradients/pull/1465/commits/6d61443f3d3b20bef16aa7fc26315640f6d7e534). Can you please fix this and we can finally merge it )

Fixed!

**Comment by BloodAxe** [comment]
Closing as stale

**Pull request closed without merging**